File: sdks/typescript/src/leader_worker.ts
// SPDX-License-Identifier: LGPL-2.1-or-later
// Copyright (C) 2025 PlexSpaces Contributors
//
// Leader-worker client for multi-node: same API surface as Rust/Python/Go SDKs.
// Virtual actors are created lazily on first message; use spawnActorOnNode
// only for non-virtual workers.

/**
 * Client for leader-worker multi-node patterns. Connect to the entry (leader) node
 * via HTTP; list worker node IDs and spawn non-virtual actors on specific nodes.
 * Virtual actors are created lazily on first message—no explicit spawn or ensure.
 */
export class LeaderWorkerClient {
  private readonly entryUrl: string;
  private readonly nodeIdToHttp: Map<string, string> = new Map();

  /**
   * @param entryHttpUrl - Base URL of the entry/leader node (e.g. "http://localhost:8092").
   */
  constructor(entryHttpUrl: string) {
    this.entryUrl = entryHttpUrl.replace(/\/$/, "");
  }

  /**
   * Derive HTTP URL from gRPC node_address (convention: HTTP = gRPC port + 1).
   */
  private static grpcPortToHttp(url: string): string {
    if (!url) return url;
    const match = url.trim().match(/:(\d+)\s*$/);
    if (match) {
      const port = parseInt(match[1], 10) + 1;
      return url.trim().replace(/:(\d+)\s*$/, `:${port}`);
    }
    return url;
  }

  /**
   * List node IDs that can run workers (peers + self). Populates internal
   * cache so spawnActorOnNode can resolve nodeId to node HTTP URL.
   *
   * @returns List of node_id strings.
   */
  async listWorkerNodeIds(
    cluster?: string | null,
    pageSize = 100,
    pageToken = ""
  ): Promise<string[]> {
    const params = new URLSearchParams({ pageSize: String(pageSize) });
    if (cluster) params.set("cluster", cluster);
    if (pageToken) params.set("pageToken", pageToken);
    const url = `${this.entryUrl}/api/v1/nodes?${params.toString()}`;
    const res = await fetch(url, {
      method: "GET",
      headers: { Accept: "application/json" },
    });
    if (!res.ok) {
      const body = await res.text();
      throw new Error(`listWorkerNodeIds failed: ${res.status} ${body}`);
    }
    const data = (await res.json()) as {
      nodes?: Array<{ nodeId?: string; node_id?: string; nodeAddress?: string; node_address?: string }>;
      nodeRegistrations?: Array<{ nodeId?: string; node_id?: string; nodeAddress?: string; node_address?: string }>;
    };
    const nodes = data.nodes ?? data.nodeRegistrations ?? [];
    const ids: string[] = [];
    for (const n of nodes) {
      const nodeId = n.nodeId ?? n.node_id ?? "";
      const nodeAddress = n.nodeAddress ?? n.node_address ?? "";
      if (nodeId) {
        ids.push(nodeId);
        if (nodeAddress) {
          this.nodeIdToHttp.set(nodeId, LeaderWorkerClient.grpcPortToHttp(nodeAddress));
        }
      }
    }
    return ids;
  }

  /**
   * Spawn a non-virtual actor on a specific node. The node must be reachable;
   * call listWorkerNodeIds first so the client can resolve nodeId to URL.
   *
   * @returns Actor ref string (e.g. "worker-ulid@node_id").
   */
  async spawnActorOnNode(
    nodeId: string,
    actorType: string,
    actorId = "",
    initialState: Uint8Array = new Uint8Array(0),
    config?: Record<string, unknown> | null,
    labels?: Record<string, string> | null
  ): Promise<string> {
    const nodeHttp = this.nodeIdToHttp.get(nodeId);
    if (!nodeHttp) {
      throw new Error(
        `unknown nodeId "${nodeId}"; call listWorkerNodeIds first`
      );
    }
    const payload: Record<string, unknown> = { actorType };
    if (actorId) payload.actorId = actorId;
    if (initialState.byteLength > 0) {
      let binary = "";
      for (let i = 0; i < initialState.length; i++) {
        binary += String.fromCharCode(initialState[i]);
      }
      payload.initialState = btoa(binary);
    }
    if (config != null) payload.config = config;
    if (labels != null) payload.labels = labels;
    const body = JSON.stringify(payload);
    const url = `${nodeHttp.replace(/\/$/, "")}/api/v1/actors/spawn`;
    const res = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json", Accept: "application/json" },
      body,
    });
    if (!res.ok) {
      const text = await res.text();
      throw new Error(`spawnActorOnNode failed: ${res.status} ${text}`);
    }
    const out = (await res.json()) as { actorRef?: string; actor_ref?: string };
    const actorRef = out.actorRef ?? out.actor_ref ?? "";
    if (!actorRef) throw new Error("spawnActorOnNode returned empty actorRef");
    return actorRef;
  }
}

/**
 * Convenience: list worker node IDs using a one-off client.
 * For multiple calls or spawnActorOnNode, use LeaderWorkerClient.
 */
export async function listWorkerNodeIds(
  entryHttpUrl: string,
  cluster?: string | null,
  pageSize = 100
): Promise<string[]> {
  const client = new LeaderWorkerClient(entryHttpUrl);
  return client.listWorkerNodeIds(cluster, pageSize);
}

File: sdks/rust/plexspaces-sdk/src/leader_worker.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2025 PlexSpaces Contributors
//
// Leader-worker helpers for multi-node: first node is leader, splits work across workers
// on same or other nodes. Prefer virtual actors: they are created lazily on first message
// receive, so the leader just sends to the canonical actor ID with no explicit ensure or spawn.
// Use spawn_actor_on_node only for non-virtual workers.

use anyhow::{Context, Result};
use std::sync::Arc;
use tonic::Request;

use plexspaces_core::{RequestContext, ServiceLocator};

/// List node IDs that can run workers (peers + self). Excludes none; filter by cluster if needed.
///
/// Use from the **leader** node to get the set of nodes for distributing work. After
/// [ConnectNodes](crate::NodeClient::connect_nodes), the registry contains peer nodes.
///
/// ## Returns
/// Vector of node_id strings (node registrations from the registry). Include the local
/// node if you want the leader to also run workers locally.
pub async fn list_worker_node_ids(
    ctx: &RequestContext,
    service_locator: Arc<dyn ServiceLocator>,
    cluster: Option<&str>,
    page_size: u32,
) -> Result<Vec<String>> {
    let registry = service_locator
        .get_node_registry()
        .await
        .ok_or_else(|| anyhow::anyhow!("NodeRegistry not registered"))?;
    let (nodes, _next_token) = registry
        .list_nodes(ctx, cluster, page_size, "")
        .await
        .map_err(|e| anyhow::anyhow!("list_nodes failed: {}", e))?;
    Ok(nodes
        .into_iter()
        .map(|n| n.node_id)
        .collect::<Vec<String>>())
}

/// Spawn a (non-virtual) actor on a specific node by calling that node's SpawnActor.
///
/// Use when you need an explicit worker on a remote node and are not using virtual
/// actors. The **leader** calls this, then sends work to the returned actor ref.
///
/// ## Arguments
/// - `node_id`: Target node ID (must be in registry after ConnectNodes).
/// - `actor_type`: Pre-deployed actor type on the target node (e.g. `"worker"`).
/// - `actor_name`: Optional logical actor name; if empty, the server generates one.
///
/// ## Returns
/// The canonical actor ref string for messaging.
#[cfg(feature = "grpc")]
pub async fn spawn_actor_on_node(
    _ctx: &RequestContext,
    service_locator: Arc<dyn ServiceLocator>,
    node_id: &str,
    actor_type: &str,
    actor_name: String,
    initial_state: Vec<u8>,
    config: Option<plexspaces_proto::actor::v1::ActorConfig>,
    labels: std::collections::HashMap<String, String>,
) -> Result<String> {
    use plexspaces_proto::actor::v1::{
        actor_service_client::ActorServiceClient, ActorSpawnSpec, SpawnActorRequest,
        SpawnActorResponse,
    };
    use plexspaces_proto::common::v1::ActorIdentity;

    let channel = service_locator
        .get_actor_service_client(node_id)
        .await
        .map_err(|e| anyhow::anyhow!("get_actor_service_client failed: {}", e))?;
    let mut client = ActorServiceClient::new(channel);

    let (role_opt, args) = plexspaces_core::legacy_spawn_init_json_to_role_and_args(&initial_state);
    let spec = ActorSpawnSpec {
        identity: Some(ActorIdentity {
            name: actor_name,
            actor_type: actor_type.to_string(),
        }),
        role: role_opt.unwrap_or_default(),
        namespace: String::new(),
        tenant_id: String::new(),
        visibility: 0,
        behavior_kind: String::new(),
        args,
        facets: vec![],
        config,
        labels,
    };

    let req = SpawnActorRequest {
        spec: Some(spec),
        namespace: String::new(),
        instances_count: 1,
    };

    let resp: SpawnActorResponse = client
        .spawn_actor(Request::new(req))
        .await
        .context("SpawnActor RPC failed")?
        .into_inner();

    if resp.actor_ref.is_empty() {
        anyhow::bail!("SpawnActor returned empty actor_ref")
    }
    Ok(resp.actor_ref)
}
